py/htm/encoders/date.py

```diff
@@ -249,12 +249,15 @@ def encode(self, inp, output=None):
     if self.seasonEncoder is not None:
       # Number the days starting at zero, intead of 1 like the datetime does.
       dayOfYear = timetuple.tm_yday - 1
+      assert(dayOfYear >= 0)
       # dayOfYear -= self.seasonEncoder.parameters.radius / 2. # Round towards the middle of the season.
       sdrs.append( self.seasonEncoder.encode(dayOfYear) )
 
     if self.dayOfWeekEncoder is not None:
-      dayOfWeek = timetuple.tm_wday + (timeOfDay) / 24.0
-      dayOfWeek -= .5 # Round towards noon, not midnight
+      hrs_ = float(timeOfDay) / 24.0 # add hours as decimal value in extension to day  
+      hrs_ = hrs_ * 0.5 # normalize hours to [0, 0.5], instead of [0, 1.0], because > 0.5 would cause rounding up to the next day
+      dayOfWeek = timetuple.tm_wday + hrs_
```

Comment:
fixed the hours, trick was to project hours only to 0..0,5, not 0..1.0, otherwise rounding skips to next day.

Reply:
There is a flaw in this solution.  At midnight the representation looses its semantic similarity with the previous day. 

```python
import datetime
import numpy as np
from htm.encoders.date import DateEncoder
from htm import SDR, Metrics
enc = DateEncoder(
            season=0,
            dayOfWeek=300,
            weekend=0,
            holiday=0,
            timeOfDay=0,
            customDays=0,
            holidays=((12, 25),))
sdr = SDR( enc.dimensions )
test_period = 24 * 3
metrics = Metrics( sdr, test_period )
now = datetime.datetime.now()
inc = datetime.timedelta(hours=1)
ovlp = []
for i in range( test_period ):
    enc.encode(now, sdr)
    now += inc
    ovlp.append( metrics.overlap.overlap )

print( metrics )

import matplotlib.pyplot as plt
plt.plot( np.arange(len(ovlp)), ovlp )
plt.xlabel("Hours after now.")
plt.ylabel("Semantic Similarity")
plt.show()
```

Reply:
```
>>> print( metrics )
```
SDR( 2100 )
    Sparsity Min/Mean/Std/Max 0.142857 / 0.142857 / 4.08069e-06 / 0.142857
    Activation Frequency Min/Mean/Std/Max 0 / 0.142856 / 0.143628 / 0.333333
    Entropy 0.743834
    Overlap Min/Mean/Std/Max 0.476667 / 0.958028 / 0.101115 / 0.98


![Figure_1](https://user-images.githubusercontent.com/19578795/62475141-75adf400-b772-11e9-808b-ff30ae0cbe86.png)

Reply:
good catch and thanks for the nice test-case! :+1: 

> here is a flaw in this solution. At midnight the representation looses its semantic similarity with the previous day.

yep, that's the price for fixing the "bias" with that solution. 
Maybe the clearest option would be to let the encoder do just what it says: `dayOfWeek` is only for days (=no hours added), and `timeOfDay` fills the latter functionality. 
Btw, on this note, I've observed that in NAB optimizations, "timeOfDay" is useless if "dayOfWeek" is used with enough bits (and that is due to the `+hrs_` part).

Reply:
ok, I see the fix now. Not sure why I thought the `*0.5` is needed.